reporting.py

```diff
@@ -13,9 +14,53 @@ def abbr_path(path, putative_prefix):
     else:
         return path
 
-def phases_str(phases):
+def phases_str(phases, max_num=None):
     '''Take a list of phase-subphase pairs and return them as a compact string'''
-    return ', '.join(['%d:%d' % ph_subph for ph_subph in phases])
+    if not max_num or len(phases) <= max_num:
+        return ' '.join(['%d:%d' % pair for pair in phases])
+    else:
+        n_first = math.floor(max_num / 2)
+        n_last = max_num - n_first
+        n_elided = len(phases) - (n_first + n_last)
+        return (' '.join(['%d:%d' % pair for pair in phases[:n_first]]) +
+            " [+%d] " % n_elided +
+            ' '.join(['%d:%d' % pair for pair in phases[n_first + n_elided:]]))
+
+
+def n_at_ph(jobs, ph):
+    return len([j for j in jobs if j.progress() == ph])
+
+def n_to_char(n):
+    if n == 0:
+        return ' '
+    elif n == 1:
+        return '.'
+    elif n == 2:
+        return ':'
+    elif n == 3:
+        return ';'
+    elif n >= 4:
+        return '!'
+    else:
+        return 'X'  # Should never be negative
```

Comment:
```suggestion
n_to_char_mapping = dict(enumerate(" .:;"))

def n_to_char(n):
    if n < 0:
        return 'X'  # Should never be negative

    return n_to_char_mapping.get(n, "!")
```
Might also like mapping to braille dots.  I feel a little inappropriate misusing those characters, but they do work nicely for 'modern' spinners and such.

Reply:
Nice.  I took a slightly different approach to this to make the "!" treated more like the other chars.

Braille dots is an interesting option I hadn't thought of.  I did figure there would be some extended charset characters I could use but wasn't sure if I wanted to step out of ASCII.